r/R/dplyr-funcs-type.R

```diff
@@ -76,6 +76,13 @@ register_bindings_type_cast <- function() {
   register_binding("as.numeric", function(x) {
     Expression$create("cast", x, options = cast_options(to_type = float64()))
   })
+  register_binding("as.Date", function(x) {
+    # base::as.Date() first converts to UTC and then extracts the date, which is
+    # why we need to go through timestamp() first - see unit tests for the real
+    # life impact of the difference between lubridate::date() and base::as.Date()
+    y <- build_expr("cast", x, options = cast_options(to_type = timestamp()))
```

Comment:
I assume that this line is depending on the default timezone for `timestamp()` to be UTC, yeah? Maybe we should be explicit about that here since to make it a little bit easier to read | reason about?

Reply:
You are right, that was done to match the default behaviour of `as.Date()`, but we actually need to support a `tz` argument.  

I think this needs a bit more work to support _all of_ (?) the following:
``` r
a <- as.POSIXct("2022-02-25 00:00:01", tz = "Europe/London")
b <- "2022-02-25 00:00:01"
c <- "2022/25/02 00:00:01"
d <- 1234
as.Date(a)
#> [1] "2022-02-25"
as.Date(a, tz = "Pacific/Marquesas")
#> [1] "2022-02-24"
as.Date(b)
#> [1] "2022-02-25"
as.Date(c, format = "%Y/%d/%m %H:%M:%S")
#> [1] "2022-02-25"
as.Date(d, origin = "1981-01-01")
#> [1] "1984-05-19"
```

<sup>Created on 2022-02-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
The though question here is how can we support date extraction / casting of a date with a different timezone, since arrow doesn't really support that at the moment (if I'm not mistaken). In the example above:
```r
as.Date(a, tz = "Pacific/Marquesas") #> [1] "2022-02-24"
```